carbonserver/render.go

```diff
@@ -565,3 +674,130 @@ func (listener *CarbonserverListener) fetchDataPB(metric string, files []string,
 	}
 	return &multi, nil
 }
+
+// Render implements Render rpc of CarbonV2 gRPC service
+func (listener *CarbonserverListener) Render(req *protov2.MultiFetchRequest, stream grpcv2.CarbonV2_RenderServer) (rpcErr error) {
+	t0 := time.Now()
+	ctx := stream.Context()
+
+	var reqPeer string
+	p, ok := peer.FromContext(stream.Context())
+	if ok {
+		reqPeer = p.Addr.String()
+	}
+
+	atomic.AddUint64(&listener.metrics.RenderRequests, 1)
+
+	accessLogger := TraceContextToZap(ctx, listener.accessLogger.With(
+		zap.String("handler", "grpc-render"),
+		zap.String("request_payload", req.String()),
+		zap.String("peer", reqPeer),
+	))
+
+	format := protoV2Format
+
+	targets := getProtoV2Targets(req)
+	tgs := getTargetNames(targets)
+	accessLogger = accessLogger.With(
+		zap.String("format", format.String()),
+		zap.Strings("targets", tgs),
+	)
+
+	logger := TraceContextToZap(ctx, listener.logger.With(
+		zap.String("handler", "grpc-render"),
+		zap.String("request_payload", req.String()),
+		zap.Strings("targets", tgs),
+		zap.String("format", format.String()),
+		zap.String("peer", reqPeer),
+	))
+
+	// Make sure we log which metric caused a panic()
+	defer func() {
+		if r := recover(); r != nil {
+			logger.Error("panic recovered",
+				zap.Stack("stack"),
+				zap.Any("error", r),
+			)
+			accessLogger.Error("fetch failed",
+				zap.Duration("runtime_seconds", time.Since(t0)),
+				zap.String("reason", "panic during serving the request"),
+				zap.Stack("stack"),
+				zap.Any("error", r),
+				zap.Int("http_code", http.StatusInternalServerError),
+			)
+			rpcErr = status.New(codes.Internal, "Panic occured, see logs for more information").Err()
+		}
+	}()
+
+	// TODO: implementing cache?
+	fromCache := false
```

Comment:
not a request for change but just an extra note: would need to pay attention to request latency and io when we started testing streaming on prod as it's one of those uncertain bits we have atm.

Reply:
Absolutely